pkg/controllers/scheduling/schedule.go

```diff
@@ -2,73 +2,175 @@ package scheduling
+
+	// Decisions returns the decisions of the schedule
+	Decisions() []clusterapiv1alpha1.ClusterDecision
+
+	// NumOfUnscheduled returns the number of unscheduled.
+	NumOfUnscheduled() int
+}
+
+type FilterResult struct {
+	Name             string   `json:"name"`
+	FilteredClusters []string `json:"filteredClusters"`
+}
+
+type PriorizeResult struct {
+	Name   string         `json:"name"`
+	Scores PrioritizeSore `json:"scores"`
+}
 
+// ScheduleResult is the result for a certain schedule.
 type scheduleResult struct {
-	feasibleClusters     int
-	scheduledDecisions   int
+	feasibleClusters     []*clusterapiv1.ManagedCluster
+	scheduledDecisions   []clusterapiv1alpha1.ClusterDecision
 	unscheduledDecisions int
+
+	filteredRecords map[string][]*clusterapiv1.ManagedCluster
+	scoreRecords    map[string]PrioritizeSore
+}
+
+type schedulerHandler struct {
+	recorder                kevents.EventRecorder
+	placementDecisionLister clusterlisterv1alpha1.PlacementDecisionLister
+	clusterClient           clusterclient.Interface
+}
+
+func NewSchedulerHandler(
+	clusterClient clusterclient.Interface, placementDecisionLister clusterlisterv1alpha1.PlacementDecisionLister, recorder kevents.EventRecorder) plugins.Handle {
+
+	return &schedulerHandler{
+		recorder:                recorder,
+		placementDecisionLister: placementDecisionLister,
+		clusterClient:           clusterClient,
+	}
+}
+
+func (s *schedulerHandler) EventRecorder() kevents.EventRecorder {
+	return s.recorder
+}
+
+func (s *schedulerHandler) DecisionLister() clusterlisterv1alpha1.PlacementDecisionLister {
+	return s.placementDecisionLister
+}
+
+func (s *schedulerHandler) ClusterClient() clusterclient.Interface {
+	return s.clusterClient
+}
+
+type pluginScheduler struct {
+	filters      []plugins.Filter
+	prioritizers []plugins.Prioritizer
 }
 
-func schedule(
+func NewPluginScheduler(handle plugins.Handle) *pluginScheduler {
+	return &pluginScheduler{
+		filters: []plugins.Filter{
+			predicate.New(handle),
+		},
+		prioritizers: []plugins.Prioritizer{
+			steady.New(handle),
+			balance.New(handle),
+		},
+	}
+}
+
+func (s *pluginScheduler) Schedule(
 	ctx context.Context,
 	placement *clusterapiv1alpha1.Placement,
 	clusters []*clusterapiv1.ManagedCluster,
-	clusterClient clusterclient.Interface,
-	placementDecisionLister clusterlisterv1alpha1.PlacementDecisionLister,
-) (*scheduleResult, error) {
-	// filter clusters with cluster predicates
-	feasibleClusters, err := matchWithClusterPredicates(placement.Spec.Predicates, clusters)
-	if err != nil {
-		return nil, err
+) (ScheduleResult, error) {
+	var err error
+	filtered := clusters
+
+	results := &scheduleResult{
+		filteredRecords: map[string][]*clusterapiv1.ManagedCluster{},
+		scoreRecords:    map[string]PrioritizeSore{},
+	}
+
+	// filter clusters
+	for _, f := range s.filters {
+		filtered, err = f.Filter(ctx, placement, filtered)
+
+		if err != nil {
+			return nil, err
+		}
+
+		results.filteredRecords[f.Name()] = filtered
```

Comment:
I think `filtered` may not be the set of clusters which meet requirements of the current filter, but the set of clusters which meet the requirements of the current filter and all previous filters in the slice of filters.

Reply:
yes, it is